win32/src/srw_lock_win32.c

```diff
@@ -285,6 +338,64 @@ void srw_lock_acquire_shared(SRW_LOCK_HANDLE handle)
     }
 }
 
+int srw_lock_try_acquire_shared(SRW_LOCK_HANDLE handle)
+{
+    int result;
+
+    if (handle == NULL)
+    {
+        /* Codes_SRS_SRW_LOCK_01_001: [ If handle is NULL then srw_lock_try_acquire_shared shall fail and return a non-zero value. ]*/
+        LogError("invalid argument SRW_LOCK_HANDLE handle=%p", handle);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        LARGE_INTEGER start = { 0 }, stop;
+
+        if (handle->doStatistics)
+        {
+            (void)QueryPerformanceCounter(&start);
+        }
+
+        /*Codes_SRS_SRW_LOCK_01_002: [ Otherwise srw_lock_try_acquire_shared shall call TryAcquireSRWLockShared. ]*/
+        if (!TryAcquireSRWLockShared(&handle->lock))
+        {
+            /* Codes_SRS_SRW_LOCK_01_003: [ If TryAcquireSRWLockShared returns FALSE, srw_lock_try_acquire_shared shall return a non-zero value. ]*/
+            result = MU_FAILURE;
```

Comment:
Looks like you need an ENUM here to differentiate between an actual error and the failure to acquire a lock. I suspect the user code will react differently in both the instances:
If the handle is NULL, the user code might error out
If the lock could not be acquired, the user is free to decide what she wants to do.

Reply:
Same comment for the other function.

---
In reply to: [461973925](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461973925) [](ancestors = 461973925)

Reply:
I think that acquiring a lock should never fail in any other way. I really don't like the old lock.h that had a return code just for taking the lock.

---
In reply to: [461974020](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461974020) [](ancestors = 461974020,461973925)

Reply:
This isn't acquiring a lock, it's trying. The underlying API itself has two return codes. I think it makes sense to be explicit about what the failure is when we don't acquire a lock.

---
In reply to: [461974555](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461974555) [](ancestors = 461974555,461974020,461973925)

Reply:
Ugh. I see where Amar is coming from: if non-zero is returned, does that mean that we did not acquire the lock - should we retry? or is it a critical failure?
Interestingly TryAcquireSRWLockExclusive also returns a BOOLEAN, so there is no difference between passing a NULL and not being able to acquire the lock.

---
In reply to: [461979716](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461979716) [](ancestors = 461979716,461974555,461974020,461973925)

Reply:
hmm, that's true. I hope they do a null check :P

---
In reply to: [461982771](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461982771) [](ancestors = 461982771,461979716,461974555,461974020,461973925)